### PR TITLE
Fix checkpoint API and improve loss scaler handling

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
This PR includes:

- Fixes checkpoint for scenarios that include either mixed precision or gradient accumulation
- Improve amp.DynamicLossScaler API by adding more validation
- Extend checkpoint test to include loss verification
- Extend basic training tests to include multiple optimizer
- Add more tests for dynamic loss scaler
- Increased number of train steps on tests to catch subtle inacuracies

### CODE REVIEW

File: orttraining/orttraining/python/experimental/amp/loss_scaler.py [hunk 1]
```diff
@@ -27,12 +29,12 @@ def loss_scale(self):
 
     @loss_scale.setter
     def loss_scale(self, loss_scale):
-        assert isinstance(loss_scale, float) and loss_scale > 0, "'loss_scale' must be a positive float"
-        self._loss_scale = loss_scale
+        assert isinstance(loss_scale, (int, float)) and loss_scale > 0, "'loss_scale' must be a positive float"
```
Comment (BowenBao) [inline orttraining/orttraining/python/experimental/amp/loss_scaler.py:32]: nit: should we remove float in the message since now that int is also allowed? `'loss_scale' must be positive`

File: orttraining/orttraining/python/experimental/amp/loss_scaler.py [hunk 2]
```diff
@@ -27,12 +29,12 @@ def loss_scale(self):
...
+        self._loss_scale = float(loss_scale)
 
     def reset(self):
         r"""Resets loss scaler internal state"""
-        raise NotImplementedError
+        self._loss_scale = float(self._initial_loss_scale)
```
Comment (BowenBao) [inline orttraining/orttraining/python/experimental/amp/loss_scaler.py:37]: nit: casting seems unnecessary

File: orttraining/orttraining/python/experimental/checkpoint.py [hunk 3]
```diff
@@ -11,11 +11,11 @@
 ################################################################################
 
 
-def experimental_state_dict(ort_trainer):
+def experimental_state_dict(ort_trainer, include_optimizer_state=True):
     if not ort_trainer._training_session:
         warnings.warn("ONNX Runtime training session is not initialized yet. "
                         "Please run train_step or eval_step at least once before calling state_dict().")
-        return {}
+        return ort_trainer._state_dict
```
Comment (BowenBao) [inline orttraining/orttraining/python/experimental/checkpoint.py:18]: what is the motivation for this change? is it possible that `_state_dict` is available when `_training_session` is `None`? In that case should we update the warning above

File: orttraining/orttraining/python/experimental/checkpoint.py [hunk 4]
```diff
@@ -29,9 +29,10 @@ def experimental_state_dict(ort_trainer):
             torch_state[n.name] = torch.from_numpy(np.array(onnx.numpy_helper.to_array(n)))
 
     # Need to remove redundant initializers and name suffices to map back to original torch state names
```
Comment (BowenBao) [inline orttraining/orttraining/python/experimental/checkpoint.py:31]: we are not removing name suffices here are we?

File: orttraining/orttraining/python/experimental/amp/loss_scaler.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/experimental/amp/loss_scaler.py:32]: The right type is float and we want to enforce the best practices. The fact that we convert int to float was just an implementation detail that adds robustness

File: orttraining/orttraining/python/experimental/amp/loss_scaler.py (hunk 2 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/experimental/amp/loss_scaler.py:37]: you are right. it was added as an extra protection in case the user sneaks into the class and manually change it

File: orttraining/orttraining/python/experimental/checkpoint.py (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/experimental/checkpoint.py:18]: That is actually a good point. The motivation was that it is possible the user calls `load_state_dict(my_state)` and before calling `train_step(input)` or `eval_step(nput)`  - which would create a training session and effectively load the state_dict into the model - they call `state_dict()` just to check whether the state was properly preloaded into the class.

It would be strange that the user laods a state into ORTTrainer and then doesnt have a way to find out whther the state was "processed" or at least "stashed"

By making this change, I don't expect any existing feature/model to break, as nothing really happens before training.

File: orttraining/orttraining/python/experimental/checkpoint.py (hunk 4 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/experimental/checkpoint.py:31]: fixed

File: orttraining/orttraining/python/experimental/checkpoint.py (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/experimental/checkpoint.py:18]: I think the warning is still necessary, as the state being returned is not complete and the user really needs to call train_step or eval_step. What would be your suggestion for rephrasing it?

File: orttraining/orttraining/python/experimental/model_desc_validation.py [hunk 5]
```diff
@@ -184,15 +184,22 @@ def _add_input_description(self, node, name, shape, dtype=None, attr_name=None):
             shape (list): shape of input description
             dtype (torch.dtype): input data type
             attr_name (str, default is None): friendly name to allow direct access to the output description
+            ignore_duplicate (bool, default is False): silently skips addition of duplicate inputs
         '''
 
         assert isinstance(name, str) and len(name) > 0, "'name' is an invalid input name"
+        not_found = True
+        if ignore_duplicate is False:
```
Comment (kit1980) [inline orttraining/orttraining/python/experimental/model_desc_validation.py:192]: `is False` is dangerous here and later, because it compares to specific object `False`, and, for example, `1 is True` is False. Probably better to re-write as  `if not ignore_duplicate:`

File: orttraining/orttraining/python/experimental/model_desc_validation.py [hunk 6]
```diff
@@ -221,20 +228,26 @@ def _add_output_description(self, node, name, shape, is_loss, dtype=None, dtype_
...
+            if id(node) == id(self.outputs):
+                not_found = all([name not in o_desc.name for o_desc in node])
+                assert not_found, f"'name' {name} already exists in the outputs description"
+                assert all([o_desc.is_loss is False for o_desc in node]) if is_loss else True,\
+                    "Only one 'is_loss' is supported at outputs description"
+            else:
+                not_found = attr_name not in dir(self)
+                assert not_found, f"'attr_name' {attr_name} already exists in the 'node'"
+        elif not_found is False:
```
Comment (kit1980) [inline orttraining/orttraining/python/experimental/model_desc_validation.py:249]: Maybe `elif not not_found:`?
Or `elif not_found == False`

File: orttraining/orttraining/python/experimental/amp/loss_scaler.py [hunk 7]
```diff
@@ -93,15 +95,16 @@ def __init__(self, automatic_update=True,
...
         self._stable_steps_count = 0
 
     def reset(self):
-        self.loss_scale = self._initial_loss_scale
+        super().reset()
         self._stable_steps_count = 0
 
     def update(self, train_step_info):
+        if not self.automatic_update:
+            return
```
Comment (kit1980) [inline orttraining/orttraining/python/experimental/amp/loss_scaler.py:106]: Maybe still return `self.loss_scale`?

**Comment by kit1980** [comment]
Not really related to this PR, but in `def _reshape_tensors(self, state_dict, fp16):` the `fp16` parameter is not used, should it?

File: orttraining/orttraining/python/experimental/checkpoint.py [hunk 8]
```diff
@@ -63,11 +64,11 @@ def experimental_load_state_dict(ort_trainer, state_dict, strict=False):
     ort_trainer._training_session.load_state(session_state, strict)
 
 
-def experimental_save_checkpoint(ort_trainer, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", checkpoint_state_dict=None):
+def experimental_save_checkpoint(ort_trainer, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", checkpoint_state_dict=None, include_optimizer_state=True):
     if checkpoint_state_dict == None:
```
Comment (kit1980) [inline orttraining/orttraining/python/experimental/checkpoint.py:68]: Should be `is None`

**Comment by thiagocrepaldi** [comment]
> _reshape_tensors

ZeRO checkpoint code was inherited from current frontend with as few as possible changes. We can fix any issue after the redesign
